### PR TITLE
Adds SLL prediction mode to parser to reduce latency

### PR DESCRIPTION
**Pull request by johnedquinn**
## Relevant Issues
- None

## Description
- Adds SLL prediction mode to parser to reduce latency
- Ran benchmarks using the PartiQLParser (suffixed with New), the SqlParser (suffixed with Old), and the modified PartiQLParser (suffixed with Experimental). Here are the results:
```
Benchmark                                     Mode  Cnt     Score     Error  Units
ParserBenchmark.nestedSelectExperimental      avgt   20   124.236 ±   5.000  us/op
ParserBenchmark.nestedSelectNew               avgt   20  2525.692 ± 116.653  us/op
ParserBenchmark.nestedSelectOld               avgt   20   145.874 ±   9.847  us/op
ParserBenchmark.simpleSelectExperimental      avgt   20    10.384 ±   0.273  us/op
ParserBenchmark.simpleSelectNew               avgt   20    16.875 ±   1.262  us/op
ParserBenchmark.simpleSelectOld               avgt   20     7.428 ±   0.803  us/op
ParserBenchmark.thirtyPlusExperimental        avgt   20    55.882 ±   5.383  us/op
ParserBenchmark.thirtyPlusNew                 avgt   20   169.454 ±   5.332  us/op
ParserBenchmark.thirtyPlusOld                 avgt   20    51.452 ±   1.560  us/op
ParserBenchmark.thirtyPredicatesExperimental  avgt   20   160.670 ±   8.865  us/op
ParserBenchmark.thirtyPredicatesNew           avgt   20   381.240 ±  27.889  us/op
ParserBenchmark.thirtyPredicatesOld           avgt   20   216.799 ±  11.786  us/op
```
- Below are the results using commit `bbf9fb2`. `new` in this table denotes the modified PartiQLParser, and `old` denotes the SqlParser.
```
Benchmark                            Mode  Cnt    Score   Error  Units
ParserBenchmark.nestedSelectNew      avgt   20  141.777 ± 4.497  us/op
ParserBenchmark.nestedSelectOld      avgt   20  134.730 ± 2.370  us/op
ParserBenchmark.simpleSelectNew      avgt   20   12.700 ± 0.198  us/op
ParserBenchmark.simpleSelectOld      avgt   20    6.553 ± 0.094  us/op
ParserBenchmark.thirtyPlusNew        avgt   20   57.363 ± 2.841  us/op
ParserBenchmark.thirtyPlusOld        avgt   20   46.299 ± 0.727  us/op
ParserBenchmark.thirtyPredicatesNew  avgt   20  167.417 ± 2.725  us/op
ParserBenchmark.thirtyPredicatesOld  avgt   20  199.705 ± 4.061  us/op
```
- Below are the results using the latest revision (commit `e9c3611`) of this PR. 
```
Benchmark                            Mode  Cnt    Score    Error  Units
ParserBenchmark.nestedSelectNew      avgt   20  137.047 ±  4.944  us/op
ParserBenchmark.nestedSelectOld      avgt   20  132.624 ±  1.988  us/op
ParserBenchmark.simpleSelectNew      avgt   20   12.571 ±  0.716  us/op
ParserBenchmark.simpleSelectOld      avgt   20    7.338 ±  0.846  us/op
ParserBenchmark.thirtyPlusNew        avgt   20   56.738 ±  2.808  us/op
ParserBenchmark.thirtyPlusOld        avgt   20   46.203 ±  0.362  us/op
ParserBenchmark.thirtyPredicatesNew  avgt   20  168.559 ±  7.644  us/op
ParserBenchmark.thirtyPredicatesOld  avgt   20  209.509 ± 10.345  us/op
```
- See branch parser-jmh for more information.

## Other Information
- Updated Unreleased Section in CHANGELOG: **NO**
  - I can add later.
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **YES**
  - In order to prove that the SLL and LL parsers are able to work together, I added [MockK](https://mockk.io/) as a test dependency.

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 1]
```diff
@@ -59,27 +68,51 @@ internal class PartiQLParser(
...
+        this.lexer.inputStream = inputStream
+        this.lexer.reset()
+        this.tokens.tokenSource = this.lexer
+        this.tokens.seek(0)
+    }
+
+    private fun resetParserSLL(query: String) {
+        resetLexer(query)
+        this.parserSLL.reset()
+        this.parserSLL.interpreter.predictionMode = PredictionMode.SLL
```
Comment (johnedquinn) [inline lang/src/org/partiql/lang/syntax/PartiQLParser.kt:105]: This is the configuration @sahays

File: lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                        from = scan(
+                            id("t", caseInsensitive(), unqualified())
+                        )
+                    )
+                )
+            }
+        )
+    }
+
+    /**
```
Comment (alancai98) [inline lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt:149]: Are there any queries or tests you've found that are syntactically correct (i.e. don't throw any syntax error when parsed with the LL parser) that result in an error when run with the SLL parser (likely due to some ambiguity error)?

**Review by alancai98 — COMMENTED** [review]
Looks good and glad that we were able to see some parsing improvements when using SLL. Left some minor comments and some clarification questions.

**Comment by alancai98** [comment]
Also was there a branch or some sample JMH benchmarking changes I could pull in to test the performance locally? If not, I could mimic your experiments locally with this branch.

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 3]
```diff
@@ -59,27 +68,51 @@ internal class PartiQLParser(
...
 
-    private fun getLexer(source: String): Lexer {
-        val inputStream = CharStreams.fromStream(source.byteInputStream(StandardCharsets.UTF_8), StandardCharsets.UTF_8)
-        val lexer = GeneratedLexer(inputStream)
+    internal fun parseUsingLL(input: String): org.partiql.lang.syntax.antlr.PartiQLParser.StatementContext {
+        resetParserLL(input)
+        return this.parserLL.statement()
+    }
+
+    private fun resetLexer(query: String) {
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply lang/src/org/partiql/lang/syntax/PartiQLParser.kt:91]: Actually -- I did find an [external discussion thread](https://groups.google.com/g/antlr-discussion/c/B2TaUFm29jE) where one of the authors of ANTLR4 said it didn't actually save time to reuse the same instances -- I'm making the changes locally to just create a Lexer on the fly. It's simpler to read.

Preliminary testing shows it has very similar results to the ones just posted.

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 4]
```diff
@@ -88,10 +121,10 @@ internal class PartiQLParser(
      * NOTE: This needs to create its own lexer. Cannot share with others due to consumption of token stream.
      */
     private fun calculateTokenToParameterOrdinals(query: String): Map<Int, Int> {
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply lang/src/org/partiql/lang/syntax/PartiQLParser.kt:123]: Ack -- moving back to creating a Lexer, so I'll leave the comment.

File: lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt (hunk 2 above)
Comment (johnedquinn) [reply lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt:149]: I have not -- however, we wouldn't want to add them to the SpyTest -- as our unit tests would rely on the internal functionality of ANTLR.

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 5]
```diff
@@ -59,27 +68,51 @@ internal class PartiQLParser(
         return visitor.visit(tree) as PartiqlAst.Statement
     }
 
-    private fun parseQuery(lexer: Lexer): ParseTree {
-        val parser = getParser(lexer)
-        return parser.statement()
+    /**
+     * To reduce latency costs, the [PartiQLParser] attempts to use [PredictionMode.SLL] and falls back to
+     * [PredictionMode.LL] if a [ParseCancellationException] is thrown by the [BailErrorStrategy].
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply lang/src/org/partiql/lang/syntax/PartiQLParser.kt:73]: Ack -- adding more info.

**Comment by johnedquinn** [comment]
> Also was there a branch or some sample JMH benchmarking changes I could pull in to test the performance locally? If not, I could mimic your experiments locally with this branch.

I updated [parser-jmh](https://github.com/partiql/partiql-lang-kotlin/tree/parser-jmh) to use the code found here.

File: lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt (hunk 2 above)
Comment (alancai98) [reply lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt:149]: Got it. May be helpful as a future step/TODO for us to know what sorts of syntactically correct queries require falling back to the LL parser with the current grammar.

Could cut a separate issue to check.

**Review by alancai98 — DISMISSED** [review]

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 6]
```diff
@@ -59,26 +61,65 @@ internal class PartiQLParser(
         return visitor.visit(tree) as PartiqlAst.Statement
     }
 
-    private fun parseQuery(lexer: Lexer): ParseTree {
-        val parser = getParser(lexer)
+    /**
+     * To reduce latency costs, the [PartiQLParser] attempts to use [PredictionMode.SLL] and falls back to
```
Comment (jpschorr) [inline lang/src/org/partiql/lang/syntax/PartiQLParser.kt:65]: Perhaps: "As performance optimization, first attempt to parse using ANTLR's fast, but less powerfull [PredictionMode.SLL], falling back to the slower [PredictionMode.LL] that is capable of parsing all valid ANTLR grammars."

File: lang/src/org/partiql/lang/syntax/PartiQLParser.kt [hunk 7]
```diff
@@ -59,26 +61,65 @@ internal class PartiQLParser(
...
+        val parser = GeneratedParser(stream)
+        parser.reset()
+        parser.interpreter.predictionMode = PredictionMode.SLL
+        parser.removeErrorListeners()
+        parser.errorHandler = BailErrorStrategy()
+        return parser
+    }
+
+    /**
+     * Creates a [GeneratedParser] that uses [PredictionMode.LL]. This is the fastest prediction mode that guarantees
```
Comment (jpschorr) [inline lang/src/org/partiql/lang/syntax/PartiQLParser.kt:113]: I know the point of the statment "This is the fastest prediction mode that guarantees correct parse results." Is supposed to be the "guarantees", but what sticks out is the "fastest".  However, the "LL" method is slower than "SLL".

I'd reword the something like "This method is capable of parsing all valid inputs for a grammar, but is slower than SLL mode.

File: lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    @Test
+    fun sllParsesCorrectly() {
+        // Arrange
+        val query00 = "SELECT a FROM t"
+
+        // Act
+        val ast = parser.parseAstStatement(query00)
+
+        // Assert
+        verify(exactly = 1) { parser.parseUsingSLL(query00) }
```
Comment (jpschorr) [inline lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt:92]: Rather than spying on the workings of the code in a test, it may be beneficial to instrument the parser directly emit an event when a fall-back is necessary. This could be useful not only for determining which of a raft of test cases correspond to 'tricky' queries, but also might allow those using PartiQL in applications/services to emit events/metrics themselves (perhaps as an aide to categorize latency classes, etc.)

**Review by jpschorr — APPROVED** [review]
I have no issues with the structure of logic of the change. But I have a few recommendations for documentation changes and additional features.

File: lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt (hunk 8 above)
Comment (johnedquinn) [reply lang/test/org/partiql/lang/syntax/PartiQLParserSpyTest.kt:92]: ACK. I'll create an issue for this.

**Review by alancai98 — APPROVED** [review]
